wdns/str_to_rdata_ubuf.c

```diff
@@ -263,70 +232,137 @@ str_to_svcparam(ubuf *u, char *keyval)
+		int af = (key == spr_ipv4hint ? AF_INET : AF_INET6);
+		size_t sz = (key == spr_ipv4hint ? 4 : 16);
+
+		tok = strtok_r(val, ",", &endp);
 		if (tok == NULL || *tok == '\0') {
 			return (wdns_res_parse_error);
 		}
 
 		do {
-			if (inet_pton(AF_INET, tok, v4buf) != 1) {
+			if (inet_pton(af, tok, buf) != 1) {
 				return (wdns_res_parse_error);
 			}
 
-			ubuf_append(u, (uint8_t *)v4buf, sizeof (v4buf));
-			val_len += sizeof(v4buf);
+			ubuf_append(u, (uint8_t *)buf, sz);
+			val_len += sz;
 		} while ((tok = strtok_r(NULL, ",", &endp)) != NULL);
 		break;
-
-	case spr_ipv6hint:
-		tok = strtok_r(NULL, ",", &endp);
-		if (tok == NULL || *tok == '\0') {
-			return (wdns_res_parse_error);
+	}
+	/*
+	 * The wire format value for "alpn" consists of at least one * "alpn-id"
+	 * prefixed by its length as a single octet, and these length-value
+	 * pairs are concatenated to form the SvcParamValue. These MUST exactly
+	 * fill the SvcParamValue; otherwise, the SvcParamValue is malformed.
+	 *
+	 * We parse arbitrary keyNNNN=val pairs in a similar way, but without
+	 * length-value paris (just one whole value).
+	 *
+	 * Note that we always wrap the two value types around double quotes in
+	 * presentation format, since escaped spaces and commas are allowed.
+	 */
+	case spr_alpn:
+	default: {
+		uint8_t c, bytes;
+		bool quoted;
+		char *src;
+		int len, i;
+		ubuf *v;
+
+		src = strdup(val);
+		len = strlen(src);
+
+		v = ubuf_new();
+		bytes = 0;
+		i = 0;
+
+		if (src[i] == '"') {
+			quoted = true;
+			i++;
+		} else {
+			quoted = false;
 		}
 
-		do {
-			if (inet_pton(AF_INET6, tok, v6buf) != 1) {
-				return (wdns_res_parse_error);
+		for (; i < len; i++) {
+			c = src[i];
+
+			/* check for escaped characters first */
+			if (c == '\\') {
+				/* shouldn't happen, but just to be safe.. */
+				if (++i >= len) {
+					return (wdns_res_parse_error);
+				}
+				c = src[i];
+				ubuf_append(v, &c, 1);
+				bytes++;
+				continue;
 			}
 
-			ubuf_append(u, (uint8_t *)v6buf, sizeof (v6buf));
-			val_len += sizeof(v6buf);
-		} while ((tok = strtok_r(NULL, ",", &endp)) != NULL);
-		break;
+			/*
+			 * We're at the end of a param if we find:
+			 *  - an unescaped, matching double quote;
+			 *  - a blank space outside of double quotes.
+			 */
+			if (c == '"') {
+				if (quoted) {
+					break;
+				} else {
+					return (wdns_res_parse_error);
+				}
+			}
+			if (c == ' ') {
+			       	if (!quoted) {
+					break;
+				}
+			}
 
-	default:
-		/*
-		 * Parse an arbitrary keyNNNN=val pair.
-		 *
-		 * The rfc doesn't specify whether these types of keys can have
-		 * multiple values or not, so for now we treat it as a single
-		 * string (that could contain multiple comma separated values).
-		*/
-		tok = strtok_r(NULL, " ", &endp);
-		if (tok == NULL || *tok == '\0') {
-			return (wdns_res_parse_error);
+			/*
+			 * An unescaped comma separates two values. Add the
+			 * current 'v' to 'u' and reset it for the next value.
```

Comment:
I thought we had converged on a reading of the spec where only alpn and keyNNN params could contain escape characters (?) I'm asking because I don't quite get the need for the first call to `rdata_from_str_string()` (the one you're suggesting, where we'd pass it the entire SvcParamValue string).

Reply:
Resolving this after an offline discussion summarized below:

The spec says that the params are treated as "character strings" in the zone file format. That said, all defined parameters other than alpn and keyNNN types have a natural and unambiguous representation which does not require quoting or escaping, so it's sufficient (for wdns) to skip `rdata_from_str_string`, as we don't need the latter's flexibility to faithfully represent any wire format value for those parameters.